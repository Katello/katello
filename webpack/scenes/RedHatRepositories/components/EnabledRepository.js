import React, { Component } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { ListView, Spinner, OverlayTrigger, Tooltip } from 'patternfly-react';
import { connect } from 'react-redux';

import RepositoryTypeIcon from './RepositoryTypeIcon';
import { setRepositoryDisabled } from '../../../redux/actions/RedHatRepositories/enabled';
import api from '../../../services/api';

class EnabledRepository extends Component {
  constructor(props) {
    super(props);

    this.state = { loading: false };
    this.setDisabled = () => {
      const {
        productId, contentId, arch, releasever, name, type,
      } = this.props;

      const disabledRepo = {
        contentId,
        productId,
        name,
        type,
        arch,
        releasever,
      };

      this.props.setRepositoryDisabled(disabledRepo);
    };

    this.disableRepository = () => {
      this.setState({ loading: true });

      const {
        productId, contentId, arch, releasever,
      } = this.props;

      const url = `/products/${productId}/repository_sets/${contentId}/disable`;

      const data = {
        id: contentId,
        product_id: productId,
        basearch: arch,
        releasever,
      };

      api
        .put(url, data)
        .then(this.setDisabled)
        .catch(() => {
          this.setState({ loading: false });
          // TODO: Add error component
        });
    };

    this.disableTooltipId = `disable-${props.id}`;
  }

  render() {
    const {
      arch, name, id, type, releasever,
    } = this.props;

    return (
      <ListView.Item
        key={id}
        actions={
          <Spinner loading={this.state.loading} inline>
            <OverlayTrigger
              overlay={<Tooltip id={this.disableTooltipId}>{__('Disable')}</Tooltip>}
              placement="bottom"
              trigger={['hover', 'focus']}
              rootClose={false}
            >
              <button
                onClick={this.disableRepository}
                style={{
                  backgroundColor: 'initial',
                  border: 'none',
                  color: '#0388ce',
                }}
              >
                <i className={cx('fa-2x', 'fa fa-minus-circle')} />
              </button>
            </OverlayTrigger>
          </Spinner>
        }
        leftContent={<RepositoryTypeIcon id={id} type={type} />}
        heading={name}
        description={`${arch} ${releasever || ''}`}
        stacked
      />
    );
  }
}

EnabledRepository.propTypes = {
  id: PropTypes.number.isRequired,
  contentId: PropTypes.number.isRequired,
  productId: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  arch: PropTypes.string.isRequired,
  releasever: PropTypes.string,
  setRepositoryDisabled: PropTypes.func.isRequired,
};

EnabledRepository.defaultProps = {
  releasever: '',
};

export default connect(null, { setRepositoryDisabled })(EnabledRepository);
