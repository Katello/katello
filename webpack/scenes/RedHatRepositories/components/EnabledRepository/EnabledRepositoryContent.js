import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { Spinner, OverlayTrigger, Tooltip } from 'patternfly-react';
import { translate as __ } from 'foremanReact/common/I18n';

const EnabledRepositoryContent = ({ loading, disableTooltipId, disableRepository }) => (
  <Spinner loading={loading} inline>
    <OverlayTrigger
      overlay={<Tooltip id={disableTooltipId}>{__('Disable')}</Tooltip>}
      placement="bottom"
      trigger={['hover', 'focus']}
      rootClose={false}
    >
      <button
        onClick={disableRepository}
        style={{
          backgroundColor: 'initial',
          border: 'none',
          color: '#0388ce',
        }}
      >
        <i className={cx('fa-2x', 'fa fa-minus-circle')} />
      </button>
    </OverlayTrigger>
  </Spinner>
);

EnabledRepositoryContent.propTypes = {
  loading: PropTypes.bool.isRequired,
  disableTooltipId: PropTypes.string.isRequired,
  disableRepository: PropTypes.func.isRequired,
};

export default EnabledRepositoryContent;
