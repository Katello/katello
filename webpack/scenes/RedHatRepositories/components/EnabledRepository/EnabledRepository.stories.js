import React from '@theforeman/vendor/react';
import { storiesOf } from '@storybook/react';
import EnabledRepository from './EnabledRepository.js';

storiesOf('RedHat Repositories Page', module).add('Enabled Repository', () => (
  <EnabledRepository
    id={638}
    name="Red Hat Enterprise Linux 6 Server Kickstart x86_64 6.8"
    releasever="6.8"
    arch="x86_64"
    type="rpm"
  />
));
