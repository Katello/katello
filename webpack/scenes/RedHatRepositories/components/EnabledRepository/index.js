import { connect } from '@theforeman/vendor/react-redux';
import { setRepositoryDisabled, loadEnabledRepos, disableRepository } from '../../../../redux/actions/RedHatRepositories/enabled';
import EnabledRepository from './EnabledRepository';

const mapStateToProps = (
  { katello: { redHatRepositories: { enabled: { pagination, search } } } },
  props,
) => ({
  ...props,
  pagination,
  search,
});

export default connect(
  mapStateToProps,
  { setRepositoryDisabled, loadEnabledRepos, disableRepository },
)(EnabledRepository);
