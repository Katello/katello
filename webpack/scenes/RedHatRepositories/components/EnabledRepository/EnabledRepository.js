import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { ListView } from 'patternfly-react';
import { sprintf, translate as __ } from 'foremanReact/common/I18n';

import RepositoryTypeIcon from '../RepositoryTypeIcon';

import EnabledRepositoryContent from './EnabledRepositoryContent';

class EnabledRepository extends Component {
  constructor(props) {
    super(props);
    this.disableTooltipId = `disable-${props.id}`;
  }

  setDisabled = () => {
    this.props.setRepositoryDisabled(this.repoForAction());
  };

  repoForAction = () => {
    const {
      productId, contentId, arch, releasever, name, type,
    } = this.props;

    return {
      contentId,
      productId,
      name,
      type,
      arch,
      releasever,
    };
  };

  reload = () => (
    this.props.loadEnabledRepos({
      ...this.props.pagination,
      search: this.props.search,
    }, true)
  );

  notifyDisabled = () => {
    window.tfm.toastNotifications.notify({
      message: sprintf(__("Repository '%(repoName)s' has been disabled."), { repoName: this.props.name }),
      type: 'success',
    });
  };

  reloadAndNotify = async (result) => {
    if (result && result.success) {
      await this.reload();
      await this.setDisabled();
      await this.notifyDisabled();
    }
  };

  disableRepository = async () => {
    const result = await this.props.disableRepository(this.repoForAction());
    this.reloadAndNotify(result);
  };

  render() {
    const {
      name, id, type, orphaned, label,
    } = this.props;

    return (
      <ListView.Item
        key={id}
        actions={
          <EnabledRepositoryContent
            loading={this.props.loading}
            disableTooltipId={this.disableTooltipId}
            disableRepository={this.disableRepository}
          />
        }
        leftContent={<RepositoryTypeIcon id={id} type={type} />}
        heading={`${name} ${orphaned ? __('(Orphaned)') : ''}`}
        description={label}
        stacked
      />
    );
  }
}

EnabledRepository.propTypes = {
  id: PropTypes.number.isRequired,
  contentId: PropTypes.number.isRequired,
  productId: PropTypes.number.isRequired,
  label: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  arch: PropTypes.string.isRequired,
  search: PropTypes.shape({
    query: PropTypes.string,
    searchList: PropTypes.string,
    filters: PropTypes.array,
  }),
  pagination: PropTypes.shape({
    page: PropTypes.number,
    perPage: PropTypes.number,
  }).isRequired,
  loading: PropTypes.bool,
  releasever: PropTypes.string,
  orphaned: PropTypes.bool,
  setRepositoryDisabled: PropTypes.func.isRequired,
  loadEnabledRepos: PropTypes.func.isRequired,
  disableRepository: PropTypes.func.isRequired,
};

EnabledRepository.defaultProps = {
  releasever: '',
  orphaned: false,
  search: {},
  loading: false,
};

export default EnabledRepository;
