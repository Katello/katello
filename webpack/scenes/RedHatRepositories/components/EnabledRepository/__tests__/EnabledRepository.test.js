import React from 'react';
import thunk from 'redux-thunk';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureMockStore from 'redux-mock-store';
import EnabledRepository from '../EnabledRepository';

const mockStore = configureMockStore([thunk]);
const store = mockStore({});

describe('Enabled Repositories Component', () => {
  let shallowWrapper;
  const mockSetRepositoryDisabled = jest.fn();
  const mockLoadEnabledRepos = jest.fn();
  const mockDisableRepository = jest.fn();
  beforeEach(() => {
    shallowWrapper = shallow(<EnabledRepository
      store={store}
      id={1}
      contentId={1}
      productId={1}
      name="foo"
      orphaned={false}
      type="foo"
      arch="foo"
      releasever="1.1.1"
      label="some label"
      pagination={{}}
      setRepositoryDisabled={mockSetRepositoryDisabled}
      loadEnabledRepos={mockLoadEnabledRepos}
      disableRepository={mockDisableRepository}
    />);
  });

  afterEach(() => {
    store.clearActions();
  });

  it('should render', async () => {
    expect(toJson(shallowWrapper)).toMatchSnapshot();
  });

  describe('class methods', () => {
    let instance;
    beforeEach(() => {
      instance = shallowWrapper.instance();
      mockSetRepositoryDisabled.mockClear();
      mockLoadEnabledRepos.mockClear();
      mockDisableRepository.mockClear();
    });
    it('setDisabled - should call this.props.setRepositoryDisabled', () => {
      instance.setDisabled();
      expect(mockSetRepositoryDisabled).toHaveBeenCalled();
    });
    it('repoForAction - should return expected object', () => {
      const expected = {
        contentId: 1,
        productId: 1,
        name: 'foo',
        type: 'foo',
        arch: 'foo',
        releasever: '1.1.1',
      };
      expect(instance.repoForAction()).toEqual(expected);
    });
    it('reload - calls this.props.loadEnabledRepos with expected data', () => {
      const expectedData = {
        search: {},
      };
      instance.reload();
      expect(mockLoadEnabledRepos).toHaveBeenCalledWith(expectedData, true);
      // this.props.loadEnabledRepos({
      //   ...this.props.pagination,
      //   search: this.props.search,
      // }, true)
    });
    it('notifyDisabled - uses window.tfm.toastNotifications object to make a toast', () => {
      // setup
      const savedTfm = window.tfm;
      window.tfm = {
        toastNotifications: {
          notify: jest.fn(),
        },
      };
      const expectedData = {
        message: "Repository 'foo' has been disabled.",
        type: 'success',
      };
      instance.notifyDisabled();
      expect(window.tfm.toastNotifications.notify).toHaveBeenCalledWith(expectedData);
      // cleanup
      window.tfm = savedTfm;
    });
    it('reloadAndNotify - calls all 3 async functions', async () => {
      const result = { success: true };
      instance.reload = jest.fn();
      instance.setDisabled = jest.fn();
      instance.notifyDisabled = jest.fn();
      await instance.reloadAndNotify(result);
      expect(instance.reload).toHaveBeenCalled();
      expect(instance.setDisabled).toHaveBeenCalled();
      expect(instance.notifyDisabled).toHaveBeenCalled();
    });
    it('disableRepository - calls both async functions', async () => {
      const repoForAction = {
        contentId: 1,
        productId: 1,
        name: 'foo',
        type: 'foo',
        arch: 'foo',
        releasever: '1.1.1',
      };
      instance.reloadAndNotify = jest.fn();
      await instance.disableRepository();
      expect(mockDisableRepository).toHaveBeenCalledWith(repoForAction);
      expect(instance.reloadAndNotify).toHaveBeenCalled();
    });
  });
});
