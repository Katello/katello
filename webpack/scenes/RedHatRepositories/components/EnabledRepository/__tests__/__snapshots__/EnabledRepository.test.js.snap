// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Enabled Repositories Component should render 1`] = `
<ListViewItem
  actions={
    <EnabledRepositoryContent
      disableRepository={[Function]}
      disableTooltipId="disable-1"
      loading={false}
    />
  }
  additionalInfo={null}
  checkboxInput={null}
  compoundExpand={false}
  compoundExpanded={false}
  description="some label"
  heading="foo "
  hideCloseIcon={false}
  initExpanded={false}
  key="1"
  leftContent={
    <RepositoryTypeIcon
      id={1}
      type="foo"
    />
  }
  onCloseCompoundExpand={[Function]}
  onExpand={[Function]}
  onExpandClose={[Function]}
  stacked={true}
/>
`;
